plugins/python/src/main/kotlin/io/paddle/plugin/python/dependencies/interpretator/AbstractInterpreterDownloader.kt

```diff
@@ -0,0 +1,95 @@
+                    ?: throw Task.ActException("Can't find an appropriate version at ${PythonPaths.PYTHON_DISTRIBUTIONS_BASE_URL} for version $userDefinedVersion")
+
+            project.terminal.info("Downloading interpreter ${matchedVersion.fullName}...")
+            val workDir =
+                project.pyLocations.interpretersDir.resolve(matchedVersion.number).toFile().also { it.mkdirs() }
+
+            val pythonDistName = "Python-${matchedVersion}"
+            val archiveDistName = "$pythonDistName.tgz"
+            val url = PythonPaths.PYTHON_DISTRIBUTIONS_BASE_URL.join(matchedVersion.number, archiveDistName).trimEnd('/')
+            val file = workDir.resolve(archiveDistName)
+
+            if (file.exists()) {
+                project.terminal.info("Found downloaded distribution archive: ${file.path}")
+            } else {
+                downloadArchive(url, file, project)
+            }
+
+            val extractDir = workDir.resolve(pythonDistName).also {
+                if (it.exists()) {
+                    it.deleteRecursively()
+                }
+                it.mkdirs()
+            }
+            project.terminal.info("Unpacking archive: ${workDir.resolve(archiveDistName)}...")
+            unpackTarGZip(workDir.resolve(archiveDistName), extractDir)
+            project.terminal.info("Unpacking finished")
+
+            installPreRequirements(project)
+
+            // TODO: support Win?
+            project.terminal.info("Installing interpreter...")
+            val localPythonDir = extractDir.resolve(PythonPaths.LOCAL_PYTHON_DIR_NAME).also { it.mkdirs() }
+            val repoDir = extractDir.resolve(pythonDistName)
+
+            project.executor.run {
+                execute("./configure", getConfigureArgs(extractDir), repoDir, env = getEnvVars(extractDir), terminal = project.terminal)
+                    .then {
+                        execute("make", emptyList(), repoDir, env = getEnvVars(extractDir), terminal = project.terminal)
+                    }.then {
+                        execute("make", listOf("install"), repoDir, env = getEnvVars(extractDir), terminal = project.terminal)
+                    }.orElseDo { code ->
+                        throw Task.ActException("Failed to install interpreter $pythonDistName. Exit code is $code")
+                    }
+            }
+            project.terminal.info("Interpreter installed to ${localPythonDir.resolve("bin").path}")
+
+            val path = localPythonDir.deepResolve("bin", matchedVersion.executableName).toPath()
+            return@runBlocking PyInterpreter(path, matchedVersion)
+        }
+
+    protected open fun installPreRequirements(project: PaddleProject) = Unit
+    protected open fun getConfigureArgs(extractDir: File) = emptyList<String>()
+    protected open fun getEnvVars(extractDir: File) = emptyMap<String, String>()
+
+    open fun findLocalInstallation(): PyInterpreter? {
```

Comment:
I don't think that this should be a method of `Downloader` class. Isn't it better to move it to `PyInterpreter` and use directly?

Reply:
I've discussed Docker and his local installations, so I bring out this method to make it tweakable